### DISCUSSION TITLE
Debian 12 Bookworm Dockerfile w/ conan

### DISCUSSION THREAD

**Discussion by athatheo** [NASA-SW-VnV/ikos#285] [Q&A]
Hi community!

Has anyone managed to use ikos on a project that handled external packages with conan package manager?
When I try it, I get the "unsupported llvm type" error.

**Comment by ivanperez-keera** [comment]
What version of LLVM are you using?

**Comment by athatheo** [comment]
> What version of LLVM are you using?

llvm-14

This is my dockerfile & I can also provide the Dockerfile to reproduce the error.

```dockerfile
RUN apt-get update && apt-get install software-properties-common --yes
RUN apt-get update && apt-get install --yes --no-install-recommends binutils cmake make python3 python3-pygments ca-certificates
RUN apt-get update && apt install --yes --no-install-recommends git build-essential checkinstall \
		python3-pytest python3-setuptools libxml2-utils libz3-dev ninja-build python3-pip
RUN pip install conan==3.3 --break-system-packages

RUN apt-add-repository "deb http://apt.llvm.org/bookworm/ llvm-toolchain-bookworm-14 main"

RUN apt-get update && apt-get install --yes --no-install-recommends libgmp-dev libboost-dev libboost-filesystem-dev \
    libboost-thread-dev libboost-test-dev libsqlite3-dev libtbb-dev \
    libz-dev libedit-dev python3-distutils python3-pip llvm-14 llvm-14-dev llvm-14-tools clang-14 libmpfr-dev python3.11-venv
RUN pip install llvmlite --break-system-packages
RUN git clone --depth 1 --branch v${IKOS_VERSION_TAG} https://github.com/NASA-SW-VnV/ikos
WORKDIR ikos
# Python >3.9 changed the function name, updating
RUN sed -i.orig 's/isAlive/is_alive/g' ./analyzer/python/ikos/analyzer.py
RUN cmake -B build -DLLVM_CONFIG_EXECUTABLE="/usr/lib/llvm-14/bin/llvm-config" -DCMAKE_INSTALL_PREFIX=/usr/ -GNinja
WORKDIR build
RUN ninja -j$(nproc)

RUN checkinstall --type=debian --pkgname=ikos --pkgversion=${IKOS_VERSION_TAG} --pkgrelease="CUSTOM" \
	--provides=ikos \
	--requires="python3 \(\>= 3.11 \),libstdc++6 \(\>= 10.2.1\),libsqlite3-0 \(\>= 3.34.1\),libboost-dev \(\>= 1.74.0\),libboost-filesystem-dev \(\>= 1.74.0\),libboost-thread-dev \(\>= 1.74.0\),python3-pygments \(\>= 2.7.1\),libgmpxx4ldbl \(\>= 6.2.1\),libtbb-dev \(\>= 2021.8.0\),zlib1g \(\>= 1.2.11\),llvm-14 \(\>= 14.0.1\),llvm-14-tools \(\>= 14.0.1\),clang-14 \(\>= 14.0.1\),libmpfr6 \(\>= 4.1.0\),file \(\>= 5.39\)" \
	--strip=yes --fstrans=no \
	--default --exclude /usr/bin/python3 ninja install
```

**Comment by athatheo** [comment]
Hey @ivanperez-keera 
Any chance you have some feedback/comment?

**Comment by ivanperez-keera** [comment]
Hi. I'm taking a look at this now to see if I can reproduce the issue. Sorry that it took some time; I had to temporarily shift attention to a different project.

**Comment by ivanperez-keera** [comment]
This is what I get in `jammy`:

```
Step 5/15 : RUN pip install conan==3.3 --break-system-packages
 ---> Running in 65e3c8e8f781

Usage:   
  pip install [options] <requirement specifier> [package-index-options] ...
  pip install [options] -r <requirements file> [package-index-options] ...
  pip install [options] [-e] <vcs project url> ...
  pip install [options] [-e] <local project path> ...
  pip install [options] <archive url/path> ...

no such option: --break-system-packages
The command '/bin/sh -c pip install conan==3.3 --break-system-packages' returned a non-zero code: 2
```

Also:

Why are you installing `llvmlite`? What is that? Can it be interfering with the installation?

Why are you replacing `isAlive`? It's no longer used.

I'm a bit confused about what you are trying to do. Are you trying to use ninja to install ikos itself? Or are you trying to use ikos in a project that uses ninja?

**Comment by ivanperez-keera** [comment]
FWIW, you have ikos available in debian non-free for modern versions. It may be easier to install the deb.

**Reply by jakoch** [reply]
It is not available in Debian (2024-12).
- not in non-free
  - https://ftp.debian.org/debian/pool/non-free/i/
  - or `echo "deb http://deb.debian.org/debian bookworm non-free non-free-firmware" | sudo tee -a /etc/apt/sources.list && apt update && apt search ikos`
- not in main (free), because of the license
  - https://packages.debian.org/search?keywords=ikos&searchon=names&suite=all&section=all
- maybe in "some-org non-free" registry?

**Reply by athatheo** [reply]
@ivanperez-keera any chance you have the source of the deb?

**Comment by ivanperez-keera** [comment]
I'm moving this to the discussions section for now. We can revisit later if we find that something needs to change in IKOS or the packages we provide. Please feel free to continue the conversation. Thanks!
